README.md
# Kafka-InfluxDB


[![Build Status](https://travis-ci.org/mre/kafka-influxdb.svg?branch=master)](https://travis-ci.org/mre/kafka-influxdb)
[![Coverage Status](https://codecov.io/gh/mre/kafka-influxdb/branch/master/graph/badge.svg)](https://codecov.io/gh/mre/kafka-influxdb)
[![Code Climate](https://codeclimate.com/github/mre/kafka-influxdb/badges/gpa.svg)](https://codeclimate.com/github/mre/kafka-influxdb)
[![PyPi Version](https://badge.fury.io/py/kafka_influxdb.svg)](https://badge.fury.io/py/kafka\_influxdb)
[![Scrutinizer](https://scrutinizer-ci.com/g/mre/kafka-influxdb/badges/quality-score.png?b=master)](https://scrutinizer-ci.com/g/mre/kafka-influxdb/?branch=master)

A Kafka consumer for InfluxDB written in Python.  
Supports InfluxDB 0.9.x and up. For InfluxDB 0.8.x support, check out the [0.3.0 tag](https://github.com/mre/kafka-influxdb/tree/v0.3.0).

## Use cases

Kafka will serve as a buffer for your metric data during high load.  
Also it's useful for sending metrics from offshore data centers with unreliable connections to your monitoring backend.

![](https://raw.githubusercontent.com/mre/kafka-influxdb/master/assets/schema-small.png)

## Quickstart

For a quick test, run kafka-influxdb inside a container alongside Kafka and InfluxDB. Some sample messages are generated automatically on startup (using kafkacat).

#### Python 2:

```
make
docker exec -it kafkainfluxdb
python -m kafka_influxdb -c config_example.yaml -s
```

#### Python 3:

```
make RUNTIME=py3
docker exec -it kafkainfluxdb
python -m kafka_influxdb -c config_example.yaml -s
```

#### PyPy 5.x

```
make RUNTIME=pypy
docker exec -it kafkainfluxdb
pypy3 -m kafka_influxdb -c config_example.yaml -s --kafka_reader=kafka_influxdb.reader.kafka_python
```

(Note that one additional flag is given: `--kafka_reader=kafka_influxdb.reader.kafka_python`. This is because PyPy is incompabile with the confluent kafka consumer which is a C-extension to librdkafka. Therefore we use the kafka\_python library here, which is compatible with PyPy but a bit slower.)

## Installation

```
pip install kafka_influxdb
kafka_influxdb -c config-example.yaml
```

## Performance


The following graph shows the number of messages/s read from Kafka for various Python versions and Kafka consumer plugins.  
This is testing against a Kafka topic with 10 partitions and five message brokers.
As you can see the best performance is achieved on Python 3 using the `-O` flag for bytecode optimization in combination with the `confluent-kafka` reader (default setup). Note that encoding and sending the data to InfluxDB might lower this maximum performance although you should still see a significant performance boost compared to logstash.

![Benchmark results](assets/benchmark.png)


## Benchmark


For a quick benchmark, you can start a complete `kafkacat -> Kafka -> kafka_influxdb -> Influxdb` setup with the following command:

    make

This will immediately start reading messages from Kafka and write them into InfluxDB. To see the output, you can use the InfluxDB cli.

    docker exec -it docker_influxdb_1 bash # Double check your container name
    influx
    use metrics
    show measurements

## Supported formats

You can write a custom encoder to support any input and output format (even fancy things like Protobuf). Look at the examples inside the `encoder` directory to get started. The following formats are officially supported:

#### Input formats

- [Collectd Graphite ASCII format](https://collectd.org/wiki/index.php/Graphite): :

```
mydatacenter.myhost.load.load.shortterm 0.45 1436357630
```

- [Collectd JSON format](https://collectd.org/wiki/index.php/JSON):

```json
[{
    "values":[
       0.6
    ],
    "dstypes":[
       "gauge"
    ],
    "dsnames":[
       "value"
    ],
    "time":1444745144.824,
    "interval":10.000,
    "host":"xx.example.internal",
    "plugin":"cpu",
    "plugin_instance":"1",
    "type":"percent",
    "type_instance":"system"
 }]
```

-   [Raw InfluxDB line protocol (e.g. for Telegraf support)](https://github.com/mre/kafka-influxdb/issues/40):

#### Output formats

- [InfluxDB 0.9.2+ line protocol format](https://influxdb.com/docs/v0.9/write_protocols/line.html): :

```
load_load_shortterm,datacenter=mydatacenter,host=myhost value="0.45" 1436357630
```

-   [InfluxDB 0.8.x JSON format](https://influxdb.com/docs/v0.8/api/reading_and_writing_data.html#writing-data-through-http) (*deprecated*)

## Configuration

Take a look at the `config-example.yaml` to find out how to create a config file.
You can overwrite the settings from the commandline. The following parameters are allowed:

| Option                                                  | Description                                                                                           |
|---------------------------------------------------------|-------------------------------------------------------------------------------------------------------|
| `-h`, `--help`                                          | Show help message and exit                                                                            |
| `--kafka_host KAFKA_HOST`                               | Hostname or IP of Kafka message broker (default: localhost)                                           |
| `--kafka_port KAFKA_PORT`                               | Port of Kafka message broker (default: 9092)                                                          |
| `--kafka_topic KAFKA_TOPIC`                             | Topic for metrics (default: my\_topic)                                                                |
| `--kafka_group KAFKA_GROUP`                             | Kafka consumer group (default: my\_group)                                                             |
| `--kafka_reader KAFKA_READER`                           | Kafka client library to use (kafka_python or confluent) (default: kafka_influxdb.reader.confluent)    |
| `--influxdb_host INFLUXDB_HOST`                         | InfluxDB hostname or IP (default: localhost)                                                          |
| `--influxdb_port INFLUXDB_PORT`                         | InfluxDB API port (default: 8086)                                                                     |
| `--influxdb_user INFLUXDB_USER`                         | InfluxDB username (default: root)                                                                     |
| `--influxdb_password INFLUXDB_PASSWORD`                 | InfluxDB password (default: root)                                                                     |
| `--influxdb_dbname INFLUXDB_DBNAME`                     | InfluxDB database to write metrics into (default: metrics)                                            |
| `--influxdb_use_ssl`                                    | Use SSL connection for InfluxDB (default: False)                                                      |
| `--influxdb_verify_ssl`                                 | Verify the SSL certificate before connecting (default: False)                                         |
| `--influxdb_timeout INFLUXDB_TIMEOUT`                   | Max number of seconds to establish a connection to InfluxDB (default: 5)                              |
| `--influxdb_use_udp`                                    | Use UDP connection for InfluxDB (default: False)                                                      |
| `--influxdb_retention_policy INFLUXDB_RETENTION_POLICY` | Retention policy for incoming metrics (default: autogen)                                              |
| `--influxdb_time_precision INFLUXDB_TIME_PRECISION`     | Precision of incoming metrics. Can be one of 's', 'm', 'ms', 'u' (default: s)                         |
| `--encoder ENCODER`                                     | Input encoder which converts an incoming message to dictionary (default: graphite)                    |
| `--buffer_size BUFFER_SIZE`                             | Maximum number of messages that will be collected before flushing to the backend (default: 1000)      |
| `-c CONFIGFILE`, `--configfile CONFIGFILE`              | Configfile path (default: None)                                                                       |
| `-s`, `--statistics`                                    | Show performance statistics (default: True)                                                           |
| `-v`, `--verbose`                                       | Set verbosity level. Increase verbosity by adding a v: -v -vv -vvv (default: 0)                       |
| `--version`                                             | Show version                                                                                          |

## Comparison with other tools

There is a Kafka input plugin and an InfluxDB output plugin for **logstash**. It supports Influxdb 0.9+. We've achieved a message throughput of around **5000 messages/second** with that setup. Check out the configuration at docker/logstash/config.conf. You can run the benchmark yourself:

```
make RUNTIME=logstash
docker exec -it logstash
logstash -f config.conf
```

Please send a Pull Request if you know of other tools that can be mentioned here.



CHANGELOG.md
## Change Log

### master (unreleased)

### v0.9.2 (2017/07/07)

* Fix hard exit when InfluxDB database could not be created (#51)

### v0.9.1 (2017/05/17)

* Add missing build requirement

### v0.9.0 (2017/05/17)

* Add timeout to flush buffer if last flush was longer than x seconds ago. (by @c0by)
* Fix pip package
* Support heapster/eventer JSON format (by @jerryjxj)

### v0.8.0 (2016/11/30)

* Update package requirements
* Integrate Confluent Kafka Python client API (with librdkafka support)
* Cleanup setuptools
* Switch from nose to pytest
* Extend travis CI configuration
* Make writing Kafka Readers much more flexible
* Improve error handling in Kafka Reader
* Improve logging messages
* General code cleanup
* Python 3.5 support
* PyPy 5.4 support
* InfluxDB 1.x support
* Deprecate Kafka sample message writer
* Deprecate collectd message writer
* Deprecate builtin benchmark functionality
* Integrate kafkacat for sample message writing
* Improve documentation
* Fix formatting of README for PyPi (thanks @di)
* Update Docker-based benchmarks
* Simplify testing by providing a Makefile-based environment
* Start kafka-influxdb with Python optimization (-O flag) by default.

### v0.7.3 (2016/10/14)

* Support for Python 3.x and PyPy 3.x
* Allow use of external encoders
* Switch from SimpleConsumer to KafkaConsumer (huge performance improvement)
* Add version flag
* Fix unit tests for Python 3
* Fix bugs of supporting mutliple fields in one sample; 
* Add initial mypy static type support
* Full changelog: https://github.com/mre/kafka-influxdb/compare/v0.7.2...v0.7.3

### v0.7.2 (2015/10/30)

* Fix values encoded as strings into Influxdb 0.9x (#9)
* Fix boolean commandline flags changed to strings (#8)
* Fix encoding of multiple measurements in one message (#12)
* Cleanup Docker setup
* Update Documentation
* PEP8 formatting
* Full changelog: https://github.com/mre/kafka-influxdb/compare/v0.7.1...v0.7.2

### v0.7.1 (2015/10/20)

* Add requirements.txt to Manifest file to fix pip install
* Full changelog: https://github.com/mre/kafka-influxdb/compare/v0.7.0...v0.7.1

### v0.7.0 (2015/10/18)

* Add support for SSL and UDP connections to InfluxDB
* Fix loading of encoders on pip install
* Update dependencies
* Refactoring, cleanup
* Full changelog: https://github.com/mre/kafka-influxdb/compare/v0.6.0...v0.7.0

### v0.6.0 (2015/10/07)

* Fix a bug where kafka-influxdb stopped consuming messages (See #5)
* Improve config handling: config file parameters get properly overwritten by cli arguments
* Logging output contains exception message
* Refactoring and cleanup
* Full changelog: https://github.com/mre/kafka-influxdb/compare/v0.5.1...v0.6.0

### v0.5.1 (2015/07/20)

* Make the application more robust by handling Kafka disconnect gracefully.
* Increase code coverage
* Fix docker-compose setup by correctly connecting kafka with zookeeper
* Minor refactoring and cleanup
* Full changelog: https://github.com/mre/kafka-influxdb/compare/v0.5.0...v0.5.1

### v0.5.0 (2015/07/14)

* Create setuptools package
* Add continuous integration
* Full changelog: https://github.com/mre/kafka-influxdb/compare/v0.4.1...v0.5.0

### v0.4.1 (2015/07/13)

* Add functionality for benchmarks and statistics
* Improve Docker test setup
* Performance improvements
* Downloads
* Full changelog: https://github.com/mre/kafka-influxdb/compare/v0.4.0...v0.4.1

### v0.4.0 (2015/07/10)

* Major rewrite which supports InfluxDB 0.9.1
* Supports custom encodings for all input and output formats
* Better error handling
* Updated documentation
* Add unit tests
* Add a license
* Update dependencies
* Add sample configurations for Collectd and InfluxDB
* Complete docker setup for easier testing
* Full changelog: https://github.com/mre/kafka-influxdb/compare/v0.3.0...v0.4.0

### v0.3.0 (2015/03/25)

* Support Influxdb versions 0.8 and 0.9
* Support json format for incoming data
* Full changelog: https://github.com/mre/kafka-influxdb/compare/1f1810c57d61a193e10b9221d83ae14bb988cb57...v0.3.0


kafka_influxdb.py
import sys
from collections import defaultdict
import argparse
import yaml
import logging
import importlib
from reader import kafka_reader
from writer import influxdb_writer
from writer import kafka_sample_writer as benchmark
import six

import traceback
import time

class KafkaInfluxDB(object):
    def __init__(self, reader, encoder, writer, config):
        """
        Setup
        """
        self.config = config
        self.reader = reader
        self.encoder = encoder
        self.writer = writer
        self.buffer = []

    def consume(self):
        """
        Run loop. Consume messages from reader, convert it to the output format and write with writer
        """
        self.init_database()

        self.start_time = time.time()
        logging.info("Listening for messages on Kafka topic %s...", self.config.kafka_topic)
        try:
            for index, raw_message in enumerate(self.reader.read(), 1):
                self.buffer.append(self.encoder.encode(raw_message))
                if index % self.config.buffer_size == 0:
                    self.flush()
        except KeyboardInterrupt:
            logging.info("Shutdown")

    def init_database(self):
        """
        Initialize the InfluxDB database if it is not already there
        """
        try:
            logging.info("Creating InfluxDB database if not exists: %s", self.config.influxdb_dbname)
            self.writer.create_database(self.config.influxdb_dbname)
        except Exception as e:
            logging.info(e)

    def flush(self):
        """ Flush values with writer """
        try:
            self.writer.write(self.buffer)
            if self.config.statistics:
                self.show_statistics()
        except Exception as e:
            logging.warning(e)
        self.buffer = []

    def show_statistics(self):
        delta = time.time() - self.start_time
        msg_per_sec = self.config.buffer_size / delta
        print("Flushing output buffer. {0:.2f} messages/s".format(msg_per_sec))
        # Reset timer
        self.start_time = time.time()

    def set_reader(self, reader):
        self.reader = reader

    def get_reader(self):
        self.reader = reader

    def set_writer(self, writer):
        self.writer = writer

    def get_writer(self):
        self.writer = writer

    def get_buffer(self):
        return self.buffer

    def get_config(self):
        return self.config

def main():
    """
    Setup consumer
    """
    config = parse_args()

    # Set verbosity level
    if config.verbose == 1:
        logging.getLogger().setLevel(logging.INFO)
    elif config.verbose > 1:
        logging.getLogger().setLevel(logging.DEBUG)

    if config.configfile:
        logging.debug("Reading config from ", config.configfile)
        values = parse_configfile(config.configfile)
        overwrite_config_values(config, values)
    else:
        logging.info("Using default configuration")

    if config.benchmark:
        logging.info("Writing sample messages for benchmark")
        bench = benchmark.KafkaSampleWriter(config)
        bench.produce_messages()

    start_consumer(config)

def start_consumer(config):
    """
    Start metrics consumer
    """
    try:
        reader = kafka_reader.KafkaReader(config.kafka_host,
                                        config.kafka_port,
                                        config.kafka_group,
                                        config.kafka_topic)
    except Exception as e:
        logging.error("The connection to Kafka can not be established.")
        sys.exit(-1)

    encoder = load_encoder(config.encoder)
    writer = influxdb_writer.InfluxDBWriter(config.influxdb_host,
                                    config.influxdb_port,
                                    config.influxdb_user,
                                    config.influxdb_password,
                                    config.influxdb_dbname,
                                    config.influxdb_retention_policy,
                                    config.influxdb_time_precision)

    client = KafkaInfluxDB(reader, encoder, writer, config)
    client.consume()

def load_encoder(encoder_name):
    """
    Creates an instance of the given encoder.
    An encoder converts a message from one format to another
    """
    encoder_module = importlib.import_module("encoder." + encoder_name)
    encoder_class = getattr(encoder_module, "Encoder")
    # Return an instance of the class
    return encoder_class()

def parse_configfile(configfile):
    """
    Read settings from file
    """
    with open(configfile) as f:
        try:
            return yaml.safe_load(f)
        except Exception as e :
            logging.fatal("Could not load default config file: ", e)
            exit(-1)

def overwrite_config_values(config, values, prefix = ""):
    """
    Overwrite default config with custom values
    """
    for key, value in six.iteritems(values):
        if type(value) == type(dict()):
            overwrite_config_values(config, value, "%s_" % key)
        elif value != '':
            setattr(config, "%s%s" % (prefix, key), value)

def parse_args():
    parser = argparse.ArgumentParser(description='A Kafka consumer for InfluxDB', formatter_class=argparse.ArgumentDefaultsHelpFormatter)
    # Parameters
    parser.add_argument('--kafka_host', type=str, default='localhost', required=False, help="Hostname or IP of Kafka message broker")
    parser.add_argument('--kafka_port', type=int, default=9092, required=False, help="Port of Kafka message broker")
    parser.add_argument('--kafka_topic', type=str, default='my_topic', required=False, help="Topic for metrics")
    parser.add_argument('--kafka_group', type=str, default='my_group', required=False, help="Kafka consumer group")
    parser.add_argument('--influxdb_host', type=str, default='localhost', required=False, help="InfluxDB hostname or IP")
    parser.add_argument('--influxdb_port', type=int, default=8086, required=False, help="InfluxDB API port")
    parser.add_argument('--influxdb_user', type=str, default='root', required=False, help="InfluxDB username")
    parser.add_argument('--influxdb_password', type=str, default='root', required=False, help="InfluxDB password")
    parser.add_argument('--influxdb_dbname', type=str, default='metrics', required=False, help="InfluXDB database to write metrics into")
    parser.add_argument('--influxdb_retention_policy', type=str, default='default', required=False, help="Retention policy for incoming metrics")
    parser.add_argument('--influxdb_time_precision', type=str, default="s", required=False, help="Precision of incoming metrics. Can be one of 's', 'm', 'ms', 'u'")
    parser.add_argument('--encoder', type=str, default='collectd_graphite_encoder', required=False, help="Input encoder which converts an incoming message to dictionary")
    parser.add_argument('--buffer_size', type=int, default=1000, required=False, help="Maximum number of messages that will be collected before flushing to the backend")
    parser.add_argument('-c', '--configfile', type=str, default=None, required=False, help="Configfile path")
    # Flags
    parser.add_argument('-s', '--statistics', default=True, action="store_true", help="Show performance statistics")
    parser.add_argument('-b', '--benchmark', default=False, action="store_true", help="Run benchmark")
    parser.add_argument('-v', '--verbose', action='count', default=0, help="Set verbosity level. Increase verbosity by adding a v: -v -vv -vvv")
    return parser.parse_args()

if __name__ == '__main__'	:
    main()
